api/include/opentelemetry/trace/propagation/b3_propagator.h

```diff
@@ -0,0 +1,278 @@
+  }
+
+private:
+  // Converts the hex numbers stored as strings into bytes stored in a buffer.
+  static void GenerateHexFromString(nostd::string_view string, int bytes, uint8_t *buf)
+  {
+    const char *str_id = string.data();
+    for (int i = 0; i < bytes; i++)
+    {
+      int tmp = HexToInt(str_id[i]);
+      if (tmp < 0)
+      {
+        for (int j = 0; j < bytes / 2; j++)
+        {
+          buf[j] = 0;
+        }
+        return;
+      }
+      if (i % 2 == 0)
+      {
+        buf[i / 2] = tmp * 16;
+      }
+      else
+      {
+        buf[i / 2] += tmp;
+      }
+    }
+  }
+
+  // Converts a single character to a corresponding integer (e.g. '1' to 1), return -1
+  // if the character is not a valid number in hex.
+  static uint8_t HexToInt(char c)
+  {
+    if (c >= '0' && c <= '9')
+    {
+      return (int)(c - '0');
+    }
+    else if (c >= 'a' && c <= 'f')
+    {
+      return (int)(c - 'a' + 10);
+    }
+    else if (c >= 'A' && c <= 'F')
+    {
+      return (int)(c - 'A' + 10);
+    }
+    else
+    {
+      return -1;
+    }
+  }
+
+  static bool IsValidHex(nostd::string_view string_view)
+  {
+    for (int i = 0; i < string_view.length(); i++)
+    {
+      if (!(string_view[i] >= '0' && string_view[i] <= '9') &&
+          !(string_view[i] >= 'a' && string_view[i] <= 'f'))
+        return false;
+    }
+    return true;
+  }
+
+  static SpanContext ExtractImpl(Getter getter, const T &carrier)
+  {
+    // all these are hex values
+    nostd::string_view trace_id;
+    nostd::string_view span_id;
+    nostd::string_view trace_flags;
+
+    // first let's try a single-header variant
+    auto singleB3Header = getter(carrier, B3CombinedHeader);
+    if (singleB3Header != "")
+    {
+      trace_id = singleB3Header.substr(0, kTraceIdHexStrLength);
+      if (singleB3Header.size() > 33)
+      {
+        span_id = singleB3Header.substr(33, kSpanIdHexStrLength);
+      }
+      if (singleB3Header.size() > 50)
+      {
+        trace_flags = singleB3Header.substr(50, kTraceFlagHexStrLength);
+      }
+    }
+    else
+    {
+      trace_id    = getter(carrier, B3TraceIdHeader);
+      span_id     = getter(carrier, B3SpanIdHeader);
+      trace_flags = getter(carrier, B3SampledHeader);
+    }
+
+    // validate ids
+    if (!IsValidHex(trace_id) || !IsValidHex(span_id) || !IsValidHex(trace_flags) ||
+        trace_id == "00000000000000000000000000000000" || span_id == "0000000000000000")
+    {
+      return SpanContext(false, false);
+    }
+
+    // now convert hex to objects
+    TraceId trace_id_obj       = GenerateTraceIdFromString(trace_id);
+    SpanId span_id_obj         = GenerateSpanIdFromString(span_id);
+    TraceFlags trace_flags_obj = GenerateTraceFlagsFromString(trace_flags);
+    return SpanContext(trace_id_obj, span_id_obj, trace_flags_obj, true);
+  }
+};
+
+// The B3Propagator class provides an interface that enables extracting and injecting
```

Comment:
Probably header description can be like :
`B3Propogator` class provides interface that enables extracting and injecting context into single header of HTTP Request.
`B3PropogatorMultiHeader` class provides interface that enables extracting and injecting context into multiple headers of HTTP Request ?